estuary-cdk/estuary_cdk/capture/common.py

```diff
@@ -510,55 +525,140 @@ def open_binding(
-    fetch_page: FetchPageFn[_BaseDocument] | None = None,
-    fetch_snapshot: FetchSnapshotFn[_BaseDocument] | None = None,
+    fetch_changes: FetchChangesFn[_BaseDocument]
+    | dict[str, FetchChangesFn[_BaseDocument]]
+    | None = None,
+    fetch_page: FetchPageFn[_BaseDocument]
+    | dict[str, FetchPageFn[_BaseDocument]]
+    | None = None,
+    fetch_snapshot: FetchSnapshotFn[_BaseDocument]
+    | dict[str, FetchSnapshotFn[_BaseDocument]]
+    | None = None,
     tombstone: _BaseDocument | None = None,
 ):
     """
     open_binding() is intended to be called by closures set as Resource.open Callables.
 
     It does 'heavy lifting' to actually capture a binding.
 
-    TODO(johnny): Separate into snapshot vs incremental tasks?
+    When fetch_changes, fetch_page, or fetch_snapshot are provided as dictionaries,
+    each function will be run as a separate subtask with its own independent state.
+    The dictionary keys are used as subtask IDs and are used to store and retrieve
+    the state for each subtask in state.inc, state.backfill, or state.snapshot.
     """
 
     prefix = ".".join(binding.resourceConfig.path())
 
     if fetch_changes:
+        if isinstance(fetch_changes, dict):
+            for subtask_id, subtask_fetch_changes in fetch_changes.items():
+
+                async def incremental_subtask(
+                    subtask_task: Task,
+                    fetch_fn=subtask_fetch_changes,
+                    subtask_id=subtask_id,
+                ):
+                    assert state.inc and isinstance(state.inc, dict)
+                    inc_state = state.inc.get(subtask_id)
+                    assert inc_state
+
+                    await _binding_incremental_task(
+                        binding,
+                        binding_index,
+                        fetch_fn,
+                        inc_state,
+                        subtask_task,
+                    )
 
-        async def closure(task: Task):
-            assert state.inc
-            await _binding_incremental_task(
-                binding, binding_index, fetch_changes, state.inc, task,
-            )
+                task.spawn_child(
+                    f"{prefix}.incremental.{subtask_id}", incremental_subtask
+                )
+        else:
+
+            async def closure(task: Task):
+                assert state.inc and not isinstance(state.inc, dict)
+                await _binding_incremental_task(
+                    binding,
+                    binding_index,
+                    fetch_changes,
+                    state.inc,
+                    task,
+                )
 
-        task.spawn_child(f"{prefix}.incremental", closure)
+            task.spawn_child(f"{prefix}.incremental", closure)
 
     if fetch_page and state.backfill:
+        if isinstance(fetch_page, dict):
+            for subtask_id, subtask_fetch_page in fetch_page.items():
+
+                async def backfill_subtask(
```

Comment:
Same question as above on the inputs to `incremental_subtask`.

Reply:
Updated. See above comment.